python/ray/air/execution/_internal/event_manager.py

```diff
@@ -163,40 +234,452 @@ def wait(
+        started_actors = 0
+        stopped_actors = 0
+
+        # We collect different kinds of futures, that we pass to ray.wait later.
+        resource_futures = []
+        stop_futures = []
+        all_task_futures = []
+
+        # ---
+        # Handle ACTOR events (only for EventType.ACTORS and EventType.ALL).
+        if event_type in [EventType.ALL, EventType.ACTORS]:
+            # We require a timeout because otherwise we'll run into a deadlock
+            # if we request more actors than the cluster can start.
+            if timeout is None:
+                raise ValueError(
+                    f"When awaiting ACTOR events, `timeout` cannot be None. This is "
+                    f"to ensure we don't run into a deadlock if not enough resources "
+                    f"are available to start actors. Got `event_type={event_type}` and "
+                    f"`timeout={timeout}`."
+                )
+
+            # First, try to stop some actors. This will not block, but processing
+            # the on_stop callbacks may take some time.
+            stopped_actors = self._try_stop_actors(max_actors=num_events)
+
+            # Get the termination futures we will later await
+            stop_futures = list(self._futures_to_actor_termination.keys())
+
+            # Then, get the resource futures (to start actors)
+            resource_futures = self._resource_manager.get_resource_futures()
+
+            # If there are no resource futures but pending actors, we should try
+            # to start some actors. This is because some resource backends do not
+            # use resource futures.
+            if not resource_futures and self._tracked_actors_to_attrs:
+                # If we have some stopped actors and `num_events` is set,
+                # make sure to subtract this so that we don't emit more events than
+                # specified.
+                if stopped_actors > 0 and num_events:
+                    max_events_left = num_events - stopped_actors
+                else:
+                    max_events_left = num_events
+
+                # Try to start some actors for up to `timeout` seconds.
+                # This will only happen if there are no other futures to await.
+                # This is to make sure that resources that become free again can
+                # be used if we use a resource backend without resource futures.
+                # We will then subtract the time this took from the timeout we pass
+                # to `ray.wait()` below. This will work because tasks are executed
+                # asynchronously, so `ray.wait(..., timeout=5)` and
+                # `time.sleep(5); ray.wait(..., timeout=0)` should yield the same
+                # ready futures.
+                started_actors, time_taken = self._try_start_actors_timeout(
+                    timeout=timeout, max_actors=max_events_left
+                )
+                timeout = max(1e-6, timeout - time_taken)
```

Comment:
see it doesn't feel like this block belongs here.
maybe these logics should be implemented on top of EventManager as callbacks of resource events? e.g.
```
check_if_resource_is_available.on_result(lambda result: Actor() if result else None)
```
something like that.

Reply:
Refactored